plugins/autowarmth.koplugin/main.lua

```diff
@@ -0,0 +1,780 @@
+    if hours then
+        hours = hours % 24 * 3600 + 0.01 -- round up, due to reduced precision in settings.reader.lua
+    end
+    return util.secondsToClock(hours, withoutSeconds)
+end
+
+function AutoWarmth:addToMainMenu(menu_items)
+    menu_items.autowarmth = {
+        text = _("Auto-warmth"),
+        checked_func = function() return self.activate ~= 0 end,
+        sub_item_table_func = function()
+            return self:getSubMenuItems()
+        end,
+    }
+end
+
+local function tidy_menu(menu, request)
+    for i = #menu, 1, -1 do
+        if menu[i].mode ~=nil then
+            if menu[i].mode ~= request then
+                table.remove(menu,i)
+            else
+                menu[i].mode = nil
+            end
+        end
+    end
+    return menu
+end
+
+local about_text = _([[Set the frontlight warmth and night mode based on a time schedule or the sun's position.
+
+There are three types of twilight:
+• Civil: You can read a newspaper.
+• Nautical: You can see the first stars.
+• Astronomical: It is really dark.
+
+Certain warmth-values can be set for every kind of twilight and sunrise, noon, sunset and midnight.
+The screen warmth is continuously adjusted to the current time.
+
+To use the sun's position, the geographical location must be entered. The calculations are very precise (deviation less than two minutes).]])
+function AutoWarmth:getSubMenuItems()
+    return {
+        {
+            text = _("About auto-warmth"),
+            callback = function()
+                UIManager:show(InfoMessage:new{
+                    text = about_text,
+                     width = math.floor(Screen:getWidth() * 0.9),
+                })
+            end,
+            keep_menu_open = true,
+            separator = true,
+        },
+        {
+            text = _("Easy mode"),
+            checked_func = function()
+                return self.easy_mode
+            end,
+            help_text = _("Easy mode disables all entries except: civil dawn, sun rise, sun set and civil dusk."),
+            callback = function(touchmenu_instance)
+                local info_text
+                if not self.easy_mode then
+                    info_text = _("Reduce entries in auto-warmth schedule?")
+                else
+                    info_text = _("Expand entries in auto-warmth schedule?")
+                end
+                UIManager:show(ConfirmBox:new{
+                    text = info_text,
+                    ok_callback = function()
+                        self.easy_mode = not self.easy_mode
+                        G_reader_settings:saveSetting("autowarmth_easy_mode", self.easy_mode)
+                        self:scheduleMidnightUpdate()
+                        -- closing menu is necessary for refreshing the menu structure
+                        if touchmenu_instance then touchmenu_instance:closeMenu() end
+                    end,
+                })
+            end,
```

Comment:
Is there any need for confirmation when you can just toggle it back and forth right there a dozen times in seconds?

Reply:
hmm, what is really necessary here is to close the menu, as a lot of the menu contents are dynamically changed. So after the user confirms, the menu is closed and when the user opens it again: Viola the new entries are here.

It would be annoying if the menu is closed without a visual feedback that something has changed.
Is there a function to change the menu contents similar to `touchmenu_instance:updateItems()`?

Reply:
Similar to? Isn't that the one? Anyway, I'm not sure I quite understand what "reduce/expand entries in auto-warmth schedule" means. Maybe just stick to "Turn on/off advanced mode?" With potentially a little explanation of what that means that's clearer than expanding entries. ^_^

Reply:
> Is there a function to change the menu contents similar to touchmenu_instance:updateItems()?

Usually, it's just that one :) In combination with `keep_menu_open = true` and `if touchmenu_instance then touchmenu_instance:updateItems() end` inside the `callback=`

Reply:
Unfortunately this does not work. If I put `if touchmenu_instance then touchmenu_instance:updateItems() end` in the callback, the menu structure does not get updated.

Reply:
Well, you might do something a bit more complex than usual then - but I don't want to look more at this plugin :)
I remember I once had something more complex in the TextEditor plugin, where I add/remove entries to the currently opened menu (to the history of open files, when opening a file).
See `TextEditor:setupWhenDoneFunc(touchmenu_instance)` if by chance you're having the same kind of problem.

Reply:
The trick is done with:
```
    touchmenu_instance.item_table = self:getSubMenuItems()
    touchmenu_instance:updateItems()
```